typing/typetexp.ml

```diff
@@ -112,10 +119,42 @@ end = struct
   (** Map indexed by type variable names. *)
   module TyVarMap = Misc.Stdlib.String.Map
 
+  let not_generic v = get_level v <> Btype.generic_level
+
+  (* These are the "global" type variables: they were in scope before
+     we started processing the current type.
+
+     INVARIANT: No types at generic_level.
+  *)
   let type_variables = ref (TyVarMap.empty : type_expr TyVarMap.t)
-  let univars        = ref ([] : (string * type_expr) list)
-  let pre_univars    = ref ([] : type_expr list)
-  let used_variables = ref (TyVarMap.empty : (type_expr * Location.t) TyVarMap.t)
+
+  (* These are variables that have been used in the currently-being-checked
+     type.
+
+     INVARIANT: No types at generic_level.
+  *)
+  let used_variables =
+    ref (TyVarMap.empty : (type_expr * Location.t) TyVarMap.t)
+
+  (* These are variables we expect to become univars (they were introduced with
+     e.g. ['a .]), but we need to make sure they don't unify first.  Why not
+     just birth them as univars? Because they might successfully unify with a
+     row variable in the ['a. < m : ty; .. > as 'a] idiom.  They are like the
+     [used_variables], but will not be globalized in [globalize_used_variables].
+
+     INVARIANT: No types at generic_level.
+  *)
+  let univars = ref ([] : (string * type_expr) list)
+  let assert_univars uvs =
```

Comment:
Thanks to your comment, I understand that `assert_univars` is the right thing to call before adding to `univars`, but I still find the name `assert_univars` confusing because the name is so unlike the implementation. Maybe `assert_not_generic`?